[Gh-1528] Configurable stack logs display

TejasRGitHub opened the pull request:
### Feature or Bugfix
- Feature


### Detail

As described in this issue - https://github.com/data-dot-all/dataall/issues/1528 , stack logs provide some key info. 

Creating a feature config called `show_stack_logs` for s3_datasets, s3_dataset_shares, and core ( for environment stack logs ) and adding logic in frontend to hide/show the logs button. 
Using decorator in backend to restrict access based on config.


### Relates
- https://github.com/data-dot-all/dataall/issues/1528

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)? N/A
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization? N/A
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features? N/A
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users? N/A
  - Have you used the least-privilege principle? How?


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

dlpzx commented on config.json:
```diff
@@ -47,11 +47,15 @@
+            "active": true,
+            "features": {
+                "show_stack_logs": "admin-only"
+            }
```
From the user experience  show_stack_logs is very similar to show_stack_logs, but implementation-wise they are pretty different. share-logs is a module specific feature that should be defined completely inside the shares module. 
- the config.json parameter should reflect the types of logs we are limiting the visibility for `show_stack_logs` vs `show_share_logs`
- `show_stack_logs` will be implemented in the Stacks core module and `show_share_logs` in the Shares module

dlpzx replied:
- Since the logs seen are share generic logs the config.json parameter should not be in `s3_datasets_shares` but in `shares_base`

TejasRGitHub replied:
Hi @dlpzx , totally agreed on the naming part. I will make the changes. 
I can put in in another config json - `shares_base`

dlpzx commented on backend/dataall/base/utils/stack_logs_utils.py:
```diff
@@ -0,0 +1,33 @@
+from dataall.base.config import config
+from dataall.base.context import get_context
+
+
```
See comment in config.json first!
We can make the utility more generic. At the moment it is very very focused in the stacks and it contains a lot of business info (where anything in `base` should be usable outside of data.all concepts, dataset, shares...)

We could add a utility similar to `dataall.base.feature_toggle_checker.is_feature_enabled`, maybe even in the same file

```
def is_feature_allowed_value(config_property: str, allowed_value: Any):
    def decorator(f):
        fn, fn_decorator = process_func(f)

        def decorated(*args, **kwargs):
            value = config.get_property(config_property)
            if value != allowed_value
                raise Exception(f'Disabled by config {config_property}')
            return fn(*args, **kwargs)

        return fn_decorator(decorated)

    return decorator
```

TejasRGitHub replied:
@dlpzx , for now, I have added the decorator function sugggested by you in the same file. Please let me know if you have a preference to keep it in another file

dlpzx commented on backend/dataall/modules/shares_base/services/share_logs_service.py:
```diff
@@ -15,6 +16,7 @@
 
 class ShareLogsService:
     @staticmethod
+    @is_stack_logs_visible(targetType='shares')
```
With the utility proposed in the comment in stack_logs_util we could rewrite this as:
```
@is_feature_enabled('modules.shares_base.features.show_share_logs', allowed_values=['enabled', 'admin_only'])
```

dlpzx replied:
Checking if a user is part of the admin group is a pretty repeated task that could be further extracted in another utility. But that is step 2 :)

TejasRGitHub replied:
I have updated the code as per above comment. Thanks once again for a details explanation and providing a better way to do this. Also this aligns with the structure used in data.all ( base, core, modules )

dlpzx commented on backend/dataall/modules/shares_base/services/share_logs_service.py:
Shouldn't we also check the permissions and the visibility of the logs in `get_share_logs`?

TejasRGitHub replied:
That's a great catch. I will add the new decorator to it.

TejasRGitHub replied:
Oh wait thought. `get_share_logs` has a check `ShareLogsService.check_view_log_permissions` and I am adding a decorator to `check_view_log_permissions` so this should take care of blocking the share logs access

TejasRGitHub replied:
I see what you meant by checking the `get_share_logs` method. I have added a check in  `check_view_log_permissions` which will take care of the allow/disallowing user

dlpzx replied:
Sounds good, I will review the current implementation, but it makes sense

dlpzx commented on backend/dataall/core/stacks/api/resolvers.py:
```diff
@@ -64,6 +65,7 @@ def resolve_task_id(context, source: Stack, **kwargs):
         return source.EcsTaskArn.split('/')[-1]
 
 
```
Move this check to the service layer (`StackService.get_stack_logs`). Despite being a bit polluted in this case, resolvers should just check if the input is valid and call the corresponding service.

dlpzx replied:
Check comments in `stack_logs_util` --->
we could use the decorator as:
```
def map_target_stack_to_config(target_type):
    if target_type == 'environment':
        return 'core.features.show_stack_logs'
    if target_type == 'dataset':
        return 'modules.s3_datasets.features.show_stack_logs'


@is_feature_allowed_values(resolve_config=map_target_stack_to_config, allowed_values=['enabled', 'admin_only'])
```

TejasRGitHub replied:
@dlpzx , thanks for all the code snippets to give me a very clear understanding. I now understand how you would want to make it generic. I really like it. 

Where do you think we should keep this `map_target_stack_to_config ` function since it is mapping specifics of dataset and environment. Should I place it here ?

dlpzx commented on frontend/src/modules/Shared/Stack/Stack.js:
```diff
@@ -96,11 +100,36 @@ export const Stack = (props) => {
       .catch((e) => dispatch({ type: SET_ERROR, error: e.message }));
   };
 
```
Instead of re-creating the logic of the backend in the frontend, we could do the same as in the `frontend/src/modules/Shares/views/ShareView.js` (for share.canViewLogs) and use a resolver in the stack object

TejasRGitHub replied:
Will do

dlpzx reviewed (changes requested):
Hi @TejasRGitHub thank you so much for the cool feature! I left some comments and suggestions!

TejasRGitHub commented:
Hi @dlpzx , Thanks a lot for the detailed review and suggestions . I have incorporated those changes and updated code. Please let me know if this looks good now

dlpzx commented on config.json:
```diff
@@ -47,7 +47,14 @@
                 "preview_data": true,
-                "glue_crawler": true
+                "glue_crawler": true,
+                "show_stack_logs": "admin-only"
```
The parameter `show_stack_logs` could be applied to any module that deploys a CloudFormation template upon module-object-creation. So in theory we could also add the `show_stack_logs` to the modules: ml_studio, notebooks and datapipelines. 
- important: what is the default value? does it change in comparison to the previous state for these modules?
- let's not forget to clearly explain this in the github pages :)

TejasRGitHub replied:
@dlpzx, we don't use the MLstudio, notebooks, pipeline module and thus I didn't know there are stack logs for them. Let me check quickly how much is the effort. 

> important: we are not adding those modules in the map_target_to_config, what would happen to their stack logs?
That's a great point. For now, I can make it return None/ or some form of incorrect string so that `is_feature_allowe...` will throw and error. Also, in the error I can specify a way to implement this

> let's not forget to clearly explain this in the github pages :) 
Yes I will add those in the readme for sure.

dlpzx replied:
Redshift does not have an associated CDK stack; this is good and in case the parameter is not set it should default to enabled

dlpzx commented on backend/dataall/base/utils/logs_utils.py:
nit: rename this file to make it unrelated to logs (potentially we can use this in absolutely any feature!)
nit2: for the same reason rename `check_if_user_allowed_view_logs` to `check_if_user_is_admin`

dlpzx commented on backend/dataall/base/utils/logs_utils.py:
```diff
@@ -0,0 +1,33 @@
+
+    return decorator
+
+
```
I think I was not very clear with the purpose of `is_feature_has_allowed_values`. It was not only to check the validity of the config values (although I really like that you added it). It was to enable or disable a feature depending on the value, same as is_feature_enabled but with string values instead of booleans. So it would be more a `is_feature_enabled_values`. And in this case the enabled_values=['enabled', 'admin-only']. We will implement the difference between `enabled` and `admin-only` separately

TejasRGitHub replied:
Yes it would be more like `is_feature_enable_on_allowed_values` ( I can rename the function according to what makes sense ). As you said, this decorator is just to check if the config has correct config values since we are departing from true/false ( boolean values ) . The logic to check based on the config value can be done in the service layer. 

Initially I was trying not to touch code in the service layer and just make use of decorator to allow/disallow some feature, but this was conflicting with how the code is structured.

dlpzx commented on backend/dataall/base/utils/logs_utils.py:
```diff
@@ -0,0 +1,33 @@
+def check_if_user_allowed_view_logs(groups, config):
+    if (config == 'admin-only' and 'DAAdministrators' not in groups) or config == 'disabled':
+        return False
+    return True
```
I would make this generic and only check if the user is admin and move the 'admin-only' logic to the stack-logs code:

def check_if_user_is_admin(groups):
    if 'DAAdministrators' not in groups:
        return False
    return True

dlpzx replied:
It is out of the scope for this PR (I'll open the issue once this one is completed) but the next step would be to assess all the places where we raise an Unauthorized error after checking if the user is admin and modify the utility accordingly (maybe it goes in the direction of another decorator that can be parametrized on config values)

dlpzx commented on backend/dataall/core/stacks/api/resolvers.py:
```diff
@@ -57,6 +58,13 @@ def resolve_events(context, source: Stack, **kwargs):
+    if not source:
+        return False
+    log_config = config.get_property(StackServiceUtils.map_target_to_config(target_type=source.stack))
+    return check_if_user_allowed_view_logs(context.groups, log_config)
```
Implement here the logic of if (config == 'admin-only' and not is_user_admin) or config == 'disabled':
        return False
    return True

dlpzx commented on backend/dataall/base/utils/logs_utils.py:
```diff
@@ -0,0 +1,33 @@
+def check_if_user_allowed_view_logs(groups, config):
+    if (config == 'admin-only' and 'DAAdministrators' not in groups) or config == 'disabled':
+        return False
+    return True
```
Unless we make this logic very generic and turn it into a decorator that disables service calls depending of if user is admin (out of scope for this PR) this code should be implemented in the stack module.

dlpzx replied:
In the same way there is a `ShareLogsService.check_view_log_permissions`

TejasRGitHub replied:
Will do.  Since the code was getting repeated , here`StackService.check_if_user_allowed_view_logs` and here `ShareLogsService.check_view_log_permissions` I thought of putting it here.

dlpzx commented on backend/dataall/core/stacks/api/resolvers.py:
```diff
@@ -57,6 +58,13 @@ def resolve_events(context, source: Stack, **kwargs):
     return json.dumps(source.events or {})
 
 
```
We can take inspiration from the resolver in shares_base/api and add guardrails for exceptions + move all logic to service.

shares_base resolver:
```
def resolve_can_view_logs(context: Context, source: ShareObject):
    try:
        return ShareLogsService.check_view_log_permissions(context.username, context.groups, source.shareUri)
    except Exception as e:
        log.error(f'Failed to resolve if user can view share logs due to: {e}')
        return False
```
equivalent for resolve_stack_visibility:
```
```
def resolve_can_view_logs(context: Context, source: Stack, **kwargs):
   if not source:
        return False
    try:
        return StackService.check_if_user_allowed_view_logs(context.username, context.groups, source.stackUri)
    except Exception as e:
        log.error(f'Failed to resolve if user can view logs due to: {e}')
        return False
```

dlpzx replied:
Coming back to the example of shares_base/resolver ---> the resolver does not contain any business logic, it just calls the service. The service `check_if_user_allowed_view_logs` is then the one decorated and contains the permission logic. The service is used by the resolver and also by other services such as `get_stack_logs` which does not need to be decorated

dlpzx replied:
class StackService:
    @is_feature_has_allowed_values(
        allowed_values=['admin-only', 'enabled', 'disabled'],
        enabled_values=['admin-only', 'enabled'],
        default_value='enabled',
        resolve_property=StackServiceUtils.map_target_to_config,
    )
    @staticmethod
    def check_if_user_allowed_view_logs(groups, target_type):
        config = config.get_property(StackServiceUtils.map_target_to_config(target_type), 'enabled')
        if config == 'admin-only' and 'DAAdministrators' not in groups:
            return False
        return True

TejasRGitHub replied:
I think I misunderstood you in your previous comment. I thought you mean to replicate logic I had `resolve_can_view_logs` to the other resolver. Now I can get what you mean. 
No exceptions ( guard rails ) , no business logic but only validations in the resolver layer.

TejasRGitHub replied:
I saw this comment - https://github.com/data-dot-all/dataall/pull/1559#discussion_r1766409253 and thought you like the exceptions in the resolver and thus implemented this way. 
I will make changes

dlpzx replied:
Almost almost :) Let's do the analysis again:

In the resolver that resolves the stack/share_logs visibility (canViewLogs value):
- call the StackLogsService or the ShareLogsService `check_view_logs_permissions`
- add try/except clause to ensure it defaults to returning False in case there in an exception - otherwise the whole ShareObject/Environment/Dataset won't be able to be returned

The `check_view_logs_permissions`:
- should be decorated with `is_feature_enabled_for_allowed_values` - raises Exceptions for disabled config values
- inside it should contain additional logic for `admin-only` and other ownership checks - returns False when not alowed - ⚠️ We should change this behavior and raise an Exception as we end up doing in `get_share_logs` for example:
```
exceptions.ResourceUnauthorized(
                username=context.username,
                action='View Share Logs',
                resource_uri=shareUri,
            )
```

Any other resolver that uses `check_view_logs` to return logs (e.g. `get_share_logs`):
- directly use check_view_logs - which should return the exception

TejasRGitHub replied:
Thanks once again @dlpzx , for the review and the explanation. I have made the updates. Please let me know if they look good

dlpzx commented on backend/dataall/modules/shares_base/services/share_logs_service.py:
```diff
@@ -15,8 +16,16 @@
+        log_config = config.get_property('modules.shares_base.features.show_share_logs')
+        if not check_if_user_allowed_view_logs(context.groups, log_config):
+            return False
+        with context.db_engine.scoped_session() as session:
```
replace line 26-27 by:
```
       if config == 'admin-only' and 'DAAdministrators' not in groups:
            return False
```

TejasRGitHub commented on backend/dataall/base/utils/logs_utils.py:
```diff
@@ -0,0 +1,33 @@
+from dataall.base.utils.decorator_utls import process_func
+
+
+def is_feature_has_allowed_values(
```
Adding `default_value` to this function

TejasRGitHub commented on backend/dataall/base/utils/logs.py:
```diff
@@ -0,0 +1,32 @@
+
+def is_feature_has_allowed_values(
+    allowed_values: List[Any],
+    default_value: Any,
```
Added this default value param

dlpzx commented on backend/dataall/base/utils/logs.py:
nit: rename file or move the code to `backend/dataall/base/feature_toggle_checker.py`

TejasRGitHub replied:
Moved to feature_toggle_checker.py

dlpzx commented on backend/dataall/modules/shares_base/services/share_logs_service.py:
```diff
@@ -43,7 +57,7 @@ def get_share_logs_query(log_stream_name):
         context = get_context()
-        if not ShareLogsService.check_view_log_permissions(context.username, context.groups, shareUri):
+        if not ShareLogsService.check_view_logs_permissions(context.username, context.groups, shareUri):
             raise exceptions.ResourceUnauthorized(
```
Small enhancement: the exception in lines 61-65 is not needed. The exception is raised in check_view_logs_permissions

TejasRGitHub replied:
There is an exception raised inside the `check_view_logs_permissions`, in case the admin-only setting is on. If it is set to enabled, it should follow the previous logic of checking of the  user belongs to dataset owner and/or stewards groups. In this case, a false is returned .

TejasRGitHub replied:
I see where the issue is . Corrected code in the next update

dlpzx commented on backend/dataall/base/utils/logs.py:
```diff
@@ -0,0 +1,32 @@
+                    f'Disabled since incorrect values in config {config_property_value}. Correct config values {allowed_values}'
+                )
+            return fn(*args, **kwargs)
+
```
I thought we had agreed to use this to enable/disable feature, not for only validation of config params. 

```
def is_feature_enabled_for_values(
    allowed_values: List[Any],
    enabled_values: List[Any],
    default_value: Any,
    resolve_property: Optional[Callable] = None,
    config_property: Optional[str] = None,
):
    def decorator(f):
        fn, fn_decorator = process_func(f)

        def decorated(*args, **kwargs):
            config_property_value = None
            if config_property is None and resolve_property is None:
                raise Exception('Config property not provided')
            if resolve_property:
                config_property_value = resolve_property(*args, **kwargs)
            if config_property:
                config_property_value = config_property
            value = config.get_property(config_property_value, default_value)
            if value not in allowed_values:
                raise Exception(
                    f'Disabled since incorrect values in config {config_property_value}. Correct config values {allowed_values}'
                )
            if value not in enabled_values:
                raise Exception(f'Disabled by config for {value=} . Enabled config values {enabled_values}')
            return fn(*args, **kwargs)

        return fn_decorator(decorated)

    return decorator
```
This removes the need to implement enabled/disabled logic in the particular shares/log/X-service

TejasRGitHub replied:
@dlpzx, I was thinking that the decorator just be used as not a feature enable / disable checker but just to check if the values are correctly put. The logic to enable and disable would then be handled in the service layers. 
I like this idea in which it checks if the values are relating to enabled/disabled config values, will make the change. I forgot / didn't think like this

dlpzx commented on backend/dataall/modules/shares_base/services/share_logs_service.py:
```diff
@@ -15,8 +16,18 @@
+        default_value='enabled',
+        config_property='modules.shares_base.features.show_share_logs',
+    )
+    def check_view_logs_permissions(username, groups, shareUri):
```
nit: we do ont need to pass the username and groups. We can get them directly from the context object `context.groups` and `context.username`

dlpzx commented on backend/dataall/modules/shares_base/api/resolvers.py:
```diff
@@ -214,7 +213,7 @@ def resolve_user_role(context: Context, source: ShareObject, **kwargs):
 
 def resolve_can_view_logs(context: Context, source: ShareObject):
-    return ShareLogsService.check_view_log_permissions(context.username, context.groups, source.shareUri)
+    return ShareLogsService.check_view_logs_permissions(context.username, context.groups, source.shareUri)
```
Why have you changed the previous? Right now getting a shareObject is going to result in exception if it is disabled due to this resolver - which means that we cannot read the rest of the info of the ShareObject
```
def resolve_can_view_logs(context: Context, source: ShareObject):
    try:
        return ShareLogsService.check_view_logs_permissions(context.username, context.groups, source.shareUri)
    except Exception as e:
        log.error(f'Failed to resolve if user can view share logs due to: {e}')
        return False
```

TejasRGitHub replied:
Updated code with the suggested changes

dlpzx reviewed (changes requested):
Hi Tejas, thanks again for all the effort put in this PR. Have a look at the summary in this comment: https://github.com/data-dot-all/dataall/pull/1559/files#r1770948390 for the changes needed :)

TejasRGitHub commented:
Hi @dlpzx , I have made the changes and tested the code locally and on aws data.all deployment. Please let me know if it looks good and can be merged

dlpzx commented on backend/dataall/core/stacks/services/stack_service.py:
```diff
@@ -213,3 +233,21 @@ def get_stack_logs(target_uri, target_type):
+        resolve_property=map_target_type_to_log_config_path,
+    )
+    def check_if_user_allowed_view_logs(target_type: str, target_uri: str, config: str):
+        context = get_context()
```
nit: instead of passing the config as argument we can directly look it up inside the check_if_user_allowed_view_logs function.

dlpzx commented:
Hi @TejasRGitHub, this is a great PR! I have left one last nit comment and resolved all other comments. I have approved it and will merge with your response.
